core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java

```diff
@@ -192,6 +192,12 @@ synchronized RecoverySourceHandler addNewRecovery(StartRecoveryRequest request,
                 if (onNewRecoveryException != null) {
                     throw onNewRecoveryException;
                 }
+                for (RecoverySourceHandler existingHandler : recoveryHandlers) {
```

Comment:
The size of this list is bound by the setting `cluster.routing.allocation.node_concurrent_outgoing_recoveries`, which defaults to 2. Note that this setting applies to the sum of all recoveries from this node, and the list here is specific for each shard. So this should not be a bottleneck ;)